### PR TITLE
Eliminate multi limit-offset nodes to emptyRelation 

### LINKED ISSUE #2822: eliminate multi limit-offset nodes to EmptyRelation if possible
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
For such query: select * from (select * from xxxx limit n offset ..) a limit  n offset m;

If outer offset(m) is greater than or equal to inner limit(n)，we can replace inner limit node with an [LogicalPlan::EmptyRelation], like this logical plan:

---Limit: skip=m, fetch=n
---- Projection: #a.column1 
------EmptyRelation

Generally, for logical plans,  if upper offset is greater than or equal to current limit, replace current limit with an [LogicalPlan::EmptyRelation].

**Describe the solution you'd like**
As above.

**Describe alternatives you've considered**
None

**Additional context**
None

### PR DESCRIPTION
**Pull request by AssHero**
# Which issue does this PR close?
Closes #2822 

 # Rationale for this change
For logical plans, if upper offset is equal or greater than current's limit, replaces with an [LogicalPlan::EmptyRelation].

For this query: select * from (select * from t1  limit 2 offset 2 ) a  order by a.column1 offset 2, we get 
the Logical Plan before:

Limit: skip=2, fetch=None                                         
-Sort: #a.column1 ASC NULLS LAST                                 
--Projection: #a.column1                                        
---Limit: skip=2, fetch=2                                      
----Projection: #t1.column1, alias=a                          
-----TableScan: t1 projection=Some([column1])

And after：

Limit: skip=2, fetch=None                                 
-Sort: #a.column1 ASC NULLS LAST                         
--Projection: #a.column1                                
---EmptyRelation                                       

# What changes are included in this PR?
add rules to eliminate multi limit-offset nodes to emptyRelation in datafusion/optimizer/src/eliminate_limit.rs

### CODE REVIEW

**Comment by alamb** [comment]
cc @ming535 - would you like to review this PR?

**Comment by ming535** [comment]
@AssHero Hi thanks for your work! I looked into this PR, and have some concerns on logical plan node like `Join` when this optimization is used and left some comments.

**Comment by AssHero** [comment]
> @AssHero Hi thanks for your work! I looked into this PR, and have some concerns on logical plan node like `Join` when this optimization is used and left some comments.

@ming535  Thanks！But I didn't see any comments here, please let me know if I miss something?

File: datafusion/optimizer/src/eliminate_limit.rs [hunk 1]
```diff
@@ -35,35 +35,80 @@ impl EliminateLimit {
...
+                .map(|plan| eliminate_limit(_optimizer, offset, plan, _optimizer_config))
+                .collect::<Result<Vec<_>>>()?;
+
+            from_plan(plan, &expr, &new_inputs)
+        }
+        // Rest: recurse and find possible LIMIT 0/Multi LIMIT OFFSET nodes
+        _ => {
+            // For those plans(projection/sort/..) which do not affect the output rows of sub-plans, we still use upper_offset;
+            // otherwise, use 0 instead.
+            let offset = match plan {
```
[the remark this answers is not in the document]
Comment (AssHero) [reply datafusion/optimizer/src/eliminate_limit.rs:86]: For join plan, the upper_offset is set to 0, and the limit node for the join is not affected.

File: datafusion/optimizer/src/eliminate_limit.rs [hunk 2]
```diff
@@ -35,35 +35,80 @@ impl EliminateLimit {
...
+    _optimizer: &EliminateLimit,
+    upper_offset: usize,
+    plan: &LogicalPlan,
+    _optimizer_config: &OptimizerConfig,
+) -> Result<LogicalPlan> {
+    match plan {
+        LogicalPlan::Limit(Limit {
+            skip: offset,
+            fetch: limit,
+            input,
```
[the remark this answers is not in the document]
Comment (AssHero) [reply datafusion/optimizer/src/eliminate_limit.rs:48]: I'll fix this, Thanks!

**Comment by AssHero** [comment]
Improvements according to the comments.

File: datafusion/optimizer/src/eliminate_limit.rs [hunk 3]
```diff
@@ -35,35 +39,98 @@ impl EliminateLimit {
...
+/// replaces LIMIT 0 with an [LogicalPlan::EmptyRelation]
+/// replaces LIMIT node whose ancestor LIMIT's skip is greater than or equal to current's fetch
+/// with an [LogicalPlan::EmptyRelation]
+fn eliminate_limit(
+    _optimizer: &EliminateLimit,
+    ancestor: &Ancestor,
+    plan: &LogicalPlan,
+    _optimizer_config: &OptimizerConfig,
+) -> Result<LogicalPlan> {
+    match plan {
```
Comment (ming535) [inline datafusion/optimizer/src/eliminate_limit.rs:60]: How about
```
match plan {
  LogicalPlan::Limit(Limit {skip, Some(fetch), input, ..}})...
}
```

Comment (ming535) [reply datafusion/optimizer/src/eliminate_limit.rs:60]: `SELECT * FROM (SELECT * FROM test_a LIMIT 5) AS test_b OFFSET 5;`
In the above query, the current plan is the one with only `fetch` (the subquery).

I thought when the current node is `Limit` with only `skip`, it should be handled the same as `Projection`, `Sort`.

File: datafusion/optimizer/src/eliminate_limit.rs [hunk 4]
```diff
@@ -128,4 +195,44 @@ mod tests {
             \n    TableScan: test";
         assert_optimized_plan_eq(&plan, expected);
     }
+
```
Comment (ming535) [inline datafusion/optimizer/src/eliminate_limit.rs:198]: By the way, can we add more tests here to handle all combination of `skip` and `fetch` between ancestor and child?

File: datafusion/optimizer/src/eliminate_limit.rs (hunk 3 above)
Comment (AssHero) [reply datafusion/optimizer/src/eliminate_limit.rs:60]: The ancestor LIMIT node is OFFSET 5 for query "SELECT * FROM (SELECT * FROM test_a LIMIT 5) AS test_b OFFSET 5". 
For this case, we should use 5 as ancestor's skip.

File: datafusion/optimizer/src/eliminate_limit.rs (hunk 4 above)
Comment (AssHero) [reply datafusion/optimizer/src/eliminate_limit.rs:198]: yes, I'll add more test cases.

File: datafusion/optimizer/src/eliminate_limit.rs (hunk 3 above)
Comment (ming535) [reply datafusion/optimizer/src/eliminate_limit.rs:60]: > Limit with only skip should be handled here. For query: SELECT * FROM (SELECT * FROM test_a LIMIT 5) AS test_b OFFSET 5;
> 
> what do you think about this?

I see, I misunderstood your code.

**Comment by AssHero** [comment]
Add more test cases.

File: datafusion/optimizer/src/eliminate_limit.rs [hunk 5]
```diff
@@ -35,35 +39,98 @@ impl EliminateLimit {
...
+                Ancestor::FromLimit { skip, .. } => skip.unwrap_or(0),
+                _ => 0,
+            };
+            // If ancestor's skip is equal or greater than current's fetch,
+            // replaces with an [LogicalPlan::EmptyRelation].
+            // For such query, the inner query(select * from xxx limit 5) should be optimized as an EmptyRelation:
+            // select * from (select * from xxx limit 5) a limit 2 offset 5;
+            match fetch {
+                Some(fetch) => {
+                    if *fetch == 0 || ancestor_skip >= *fetch {
```
Comment (alamb) [inline datafusion/optimizer/src/eliminate_limit.rs:74]: 👍

File: datafusion/optimizer/src/eliminate_limit.rs [hunk 6]
```diff
@@ -35,35 +39,98 @@ impl EliminateLimit {
...
+
+            from_plan(plan, &expr, &new_inputs)
+        }
+        // Rest: recurse and find possible LIMIT 0/Multi LIMIT OFFSET nodes
+        _ => {
+            // For those plans(projection/sort/..) which do not affect the output rows of sub-plans, we still use ancestor;
+            // otherwise, use NotRelevant instead.
+            let ancestor = match plan {
+                LogicalPlan::Projection { .. } | LogicalPlan::Sort { .. } => ancestor,
+                _ => &Ancestor::NotRelevant,
```
Comment (alamb) [inline datafusion/optimizer/src/eliminate_limit.rs:108]: I wonder if `Ancester::Unknown` would make the intent clearer compared to `Ancester::NotRelevant`?

File: datafusion/optimizer/src/eliminate_limit.rs [hunk 7]
```diff
@@ -128,4 +200,114 @@ mod tests {
...
+            .unwrap();
+
+        let expected = "Limit: skip=2, fetch=1\
+            \n  Sort: #test.a\
+            \n    EmptyRelation";
+        assert_optimized_plan_eq(&plan, expected);
+    }
+
+    #[test]
+    fn limit_fetch_with_ancestor_limit_fetch() {
```
Comment (alamb) [inline datafusion/optimizer/src/eliminate_limit.rs:245]: 👍

**Review by alamb — APPROVED** [review]
Thank you @AssHero for the code and @ming535  for the review  -- I think the code looks good to me.

I wonder if you have seen queries that produce 0 rows much in practice (as in I wonder how common will this optimization kick in / be useful)

File: datafusion/optimizer/src/eliminate_limit.rs (hunk 6 above)
Comment (AssHero) [reply datafusion/optimizer/src/eliminate_limit.rs:108]: > I wonder if `Ancester::Unknown` would make the intent clearer compared to `Ancester::NotRelevant`?

This is consistent with limit_push_down.

**Comment by alamb** [comment]
Thanks again all!

**Pull request merged**